File: MiniDOMBase/DMD/Anagrafica/Fonti/CFonteExternalStats.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Databases



Partial Public Class Anagrafica

    Public Class CFonteExternalStats
        Implements IComparable, XML.IDMDXMLSerializable

        Public Fonte As CFonte               'Fonte
        Public IDAnnuncio As String          'Stringa che identifica la fonte sui siti esterni
        Public Conteggio As Integer          'Numero di visualizzazioni totali

        Public DI As Date?       'Data della prima visualizzazione
        Public DF As Date?       'Data dell'ultima visualizzazione
        Public DV_AVE As Integer             'Visualizzazioni per giorno (medie)
        Public DV_MIN As Integer             'Visualizzazioni per giorno (minimo)
        Public DV_MAX As Integer             'Visualizzazioni per giorno (massimo)
        Public DV_STD As Integer             'Visualizzazioni per giorno (deviazione standard)

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
            Me.Fonte = Nothing
            Me.Conteggio = 0
            Me.DI = Nothing
            Me.DF = Nothing
        End Sub

        Public Sub New(ByVal fonte As CFonte)
            Me.New()
            If (fonte Is Nothing) Then Throw New ArgumentNullException("fonte")
            Me.Fonte = fonte
        End Sub

        Public Sub New(ByVal idAnnuncio As String)
            Me.New()
            idAnnuncio = Trim(idAnnuncio)
            If (idAnnuncio = "") Then Throw New ArgumentNullException("idAnnuncio")
            Me.IDAnnuncio = idAnnuncio
        End Sub

        Private Function CompareTo(obj As Object) As Integer Implements IComparable.CompareTo
            Dim tmp As CFonteExternalStats = obj
            Dim ret As Integer = tmp.Conteggio - Me.Conteggio
            If (ret = 0) Then ret = Sistema.Strings.Compare(Me.IDAnnuncio, tmp.IDAnnuncio)
            Return ret
        End Function

        Private Sub SetFieldInternal(fieldName As String, fieldValue As Object) Implements XML.IDMDXMLSerializable.SetFieldInternal
            Select Case fieldName
                Case "Fonte"
                Case "IDAnnuncio" : Me.IDAnnuncio = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Conteggio" : Me.Conteggio = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "DI" : Me.DI = XML.Utils.Serializer.DeserializeDate(fieldValue)
                Case "DF" : Me.DF = XML.Utils.Serializer.DeserializeDate(fieldValue)
                Case "DV_AVE" : Me.DV_AVE = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "DV_MIN" : Me.DV_MIN = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "DV_MAX" : Me.DV_MAX = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "DV_STD" : Me.DV_STD = XML.Utils.Serializer.DeserializeInteger(fieldValue)
            End Select
        End Sub

        Private Sub XMLSerialize(writer As XML.XMLWriter) Implements XML.IDMDXMLSerializable.XMLSerialize
            writer.WriteAttribute("Fonte", GetID(Me.Fonte))
            writer.WriteAttribute("IDAnnuncio", Me.IDAnnuncio)
            writer.WriteAttribute("Conteggio", Me.Conteggio)
            writer.WriteAttribute("DI", Me.DI)
            writer.WriteAttribute("DF", Me.DF)
            writer.WriteAttribute("DV_AVE", Me.DV_AVE)
            writer.WriteAttribute("DV_MIN", Me.DV_MIN)
            writer.WriteAttribute("DV_MAX", Me.DV_MAX)
            writer.WriteAttribute("DV_STD", Me.DV_STD)
        End Sub

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub
    End Class



End Class

File: MiniDOMBase/DMD/Anagrafica/Fonti/Stats/CFonteExternalStats.cs
using System;
using DMD.Databases;
using DMD.XML;
using static minidom.Sistema;

namespace minidom
{
    public partial class Anagrafica
    {

        /// <summary>
        /// Statistiche relative ad una fonte
        /// </summary>
        [Serializable]
        public class CFonteExternalStats
            : IComparable, IDMDXMLSerializable, IComparable<CFonteExternalStats>
        {

            /// <summary>
            /// Fonte
            /// </summary>
            public IFonte Fonte;               

            /// <summary>
            /// Stringa che identifica la fonte sui siti esterni
            /// </summary>
            public string IDAnnuncio;

            /// <summary>
            /// Numero di visualizzazioni totali
            /// </summary>
            public int Conteggio;

            /// <summary>
            /// Data della prima visualizzazione
            /// </summary>
            public DateTime? DI;

            /// <summary>
            /// Data dell'ultima visualizzazione
            /// </summary>
            public DateTime? DF;

            /// <summary>
            /// Visualizzazioni per giorno (medie)
            /// </summary>
            public int DV_AVE;

            /// <summary>
            /// Visualizzazioni per giorno (minimo)
            /// </summary>
            public int DV_MIN;

            /// <summary>
            /// Visualizzazioni per giorno (massimo)
            /// </summary>
            public int DV_MAX;

            /// <summary>
            /// Visualizzazioni per giorno (deviazione standard)
            /// </summary>
            public int DV_STD;             

            /// <summary>
            /// Costruttore
            /// </summary>
            public CFonteExternalStats()
            {
                ////DMDObject.IncreaseCounter(this);
                Fonte = null;
                Conteggio = 0;
                DI = default;
                DF = default;
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="fonte"></param>
            public CFonteExternalStats(IFonte fonte) : this()
            {
                if (fonte is null)
                    throw new ArgumentNullException("fonte");
                Fonte = fonte;
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="idAnnuncio"></param>
            public CFonteExternalStats(string idAnnuncio) : this()
            {
                idAnnuncio = DMD.Strings.Trim(idAnnuncio);
                if (string.IsNullOrEmpty(idAnnuncio))
                    throw new ArgumentNullException("idAnnuncio");
                IDAnnuncio = idAnnuncio;
            }

            /// <summary>
            /// Compara due oggetti
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public int CompareTo(CFonteExternalStats obj)
            {
                int ret = obj.Conteggio - this.Conteggio;
                if (ret == 0) ret = DMD.Strings.Compare(IDAnnuncio, obj.IDAnnuncio, true);
                return ret;
            }
            int IComparable.CompareTo(object obj)
            {
                return this.CompareTo((CFonteExternalStats)obj);
            }

            /// <summary>
            /// Deserializzazione xml
            /// </summary>
            /// <param name="fieldName"></param>
            /// <param name="fieldValue"></param>
            void IDMDXMLSerializable.SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "Fonte":
                        {
                            break;
                        }

                    case "IDAnnuncio":
                        {
                            IDAnnuncio = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Conteggio":
                        {
                            Conteggio = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "DI":
                        {
                            DI = DMD.XML.Utils.Serializer.DeserializeDate(fieldValue);
                            break;
                        }

                    case "DF":
                        {
                            DF = DMD.XML.Utils.Serializer.DeserializeDate(fieldValue);
                            break;
                        }

                    case "DV_AVE":
                        {
                            DV_AVE = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "DV_MIN":
                        {
                            DV_MIN = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "DV_MAX":
                        {
                            DV_MAX = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "DV_STD":
                        {
                            DV_STD = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }
                }
            }

            /// <summary>
            /// Serializzazione xml
            /// </summary>
            /// <param name="writer"></param>
            void IDMDXMLSerializable.XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("Fonte", DBUtils.GetID(this.Fonte, 0));
                writer.WriteAttribute("IDAnnuncio", IDAnnuncio);
                writer.WriteAttribute("Conteggio", Conteggio);
                writer.WriteAttribute("DI", DI);
                writer.WriteAttribute("DF", DF);
                writer.WriteAttribute("DV_AVE", DV_AVE);
                writer.WriteAttribute("DV_MIN", DV_MIN);
                writer.WriteAttribute("DV_MAX", DV_MAX);
                writer.WriteAttribute("DV_STD", DV_STD);
            }


            /// <summary>
            /// Distruttore
            /// </summary>
            ~CFonteExternalStats()
            {
                ////DMDObject.DecreaseCounter(this);
            }
        }
    }
}